lib/web_ui/lib/src/engine/canvaskit/overlay_scene_optimizer.dart

```diff
@@ -0,0 +1,200 @@
+///   * Platform views: being HTML content that is to be composited along with
+///     the Flutter content.
+class Rendering {
+  final List<RenderingEntity> entities = <RenderingEntity>[];
+
+  void add(RenderingEntity entity) {
+    entities.add(entity);
+  }
+
+  /// Returns [true] if this is equibalent to [other] for use in rendering.
+  bool equalsForRendering(Rendering other) {
+    if (other.entities.length != entities.length) {
+      return false;
+    }
+    for (int i = 0; i < entities.length; i++) {
+      if (!entities[i].equalsForRendering(other.entities[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  List<RenderingRenderCanvas> get canvases =>
+      entities.whereType<RenderingRenderCanvas>().toList();
+}
+
+/// An element of a [Rendering]. Either a render canvas or a platform view.
+sealed class RenderingEntity {
+  /// Returns [true] if this entity is equal to [other] for use in a rendering.
+  ///
+  /// For example, all [RenderingRenderCanvas] objects are equal to each other
+  /// for purposes of rendering since any canvas in that place in the rendering
+  /// will be equivalent. Platform views are only equal if they are for the same
+  /// view id.
+  bool equalsForRendering(RenderingEntity other);
+}
+
+class RenderingRenderCanvas extends RenderingEntity {
+  RenderingRenderCanvas();
+
+  /// The [pictures] which should be rendered in this canvas.
+  final List<CkPicture> pictures = <CkPicture>[];
+
+  /// The [DisplayCanvas] that will be used to display [pictures].
+  ///
+  /// This is set by the view embedder.
+  DisplayCanvas? displayCanvas;
+
+  /// Adds the [picture] to the pictures that should be rendered in this canvas.
+  void add(CkPicture picture) {
+    pictures.add(picture);
+  }
+
+  @override
+  bool equalsForRendering(RenderingEntity other) {
+    return other is RenderingRenderCanvas;
+  }
+
+  @override
+  String toString() {
+    return '$RenderingRenderCanvas(${pictures.length} pictures)';
+  }
+}
+
+/// A platform view to be rendered.
+class RenderingPlatformView extends RenderingEntity {
+  RenderingPlatformView(this.viewId);
+
+  /// The [viewId] of the platform view to render.
+  final int viewId;
+
+  @override
+  bool equalsForRendering(RenderingEntity other) {
+    return other is RenderingPlatformView && other.viewId == viewId;
+  }
+
+  @override
+  String toString() {
+    return '$RenderingPlatformView($viewId)';
+  }
+}
+
+// Computes the bounds of the platform view from its associated parameters.
+@visibleForTesting
+ui.Rect computePlatformViewBounds(EmbeddedViewParams params) {
+  ui.Rect currentClipBounds = ui.Rect.largest;
+
+  // The transforms are in logical pixels, but we want to compute physical
+  // pixel bounds, so transform by the device pixel ratio.
+  final double scale = EngineFlutterDisplay.instance.devicePixelRatio;
+  Matrix4 currentTransform = Matrix4.diagonal3Values(scale, scale, 1);
```

Comment:
Correct me if I'm wrong on my thinking on this, but it seems like this should go at the end of the transformation chain rather than the start. Since, as your comment above states, the transforms are in logical pixels, it seems like multiplying by the scaling factor *first* will actually treat all these transforms as if they are in physical pixels. For example, if you had a 3:1 physical-to-logical pixel ratio, and one of your mutators is a transform that translates the view by 1 "unit", if you multiply by the scale factor first and then apply the mutator, it seems like the mutator would only actually translate the picture by 1 physical pixel or 1/3 of a logical pixel, but it probably intends to move it over 1 full logical pixel (or 3 physical pixels). I could be misunderstanding the situation though.

Reply:
Good catch. The way this was written was incorrect before. The transforms, clips, and platform view width and offset _are_ all in physical pixels. There is a transform that should come at the beginning of the mutator list which applies the devicePixelRatio skill. The bug before was that I was iterating over the mutators in the reverse order that I should have, so I applied the scaling myself, then un-applied it by dividing by the scale at the end. I fixed it properly by iterating over the mutators in the correct order now